src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= ActiveMemory.Length, $"MultiArrayBuffer.Discard: Expected byteCount={byteCount} <= {ActiveMemory.Length}");
+
+            int oldStartBlock = _activeStart / BlockSize;
+            _activeStart += byteCount;
+            int newStartBlock = _activeStart / BlockSize;
+
+            while (oldStartBlock < newStartBlock)
+            {
+                Debug.Assert(_blocks is not null);
+                Debug.Assert(_blocks[oldStartBlock] is not null, $"Discard: oldStartBlock is null?? byteCount={byteCount}, _activeStart={_activeStart}, oldStartBlock={oldStartBlock}, newStartBlock={newStartBlock}");
+
+                ArrayPool<byte>.Shared.Return(_blocks[oldStartBlock]!);
+                _blocks[oldStartBlock] = null;
+
+                oldStartBlock++;
+            }
+
+            if (_activeStart == _availableStart)
+            {
+                // Small optimization to restart at the beginning of the current block, since we have no active bytes.
+                // Note, we don't try to release buffers in this case. Maybe we should. But if we did,
+                // we'd need handle the fact that there could be more than one here, if we previously grew the buffer enough for that.
+
+                _activeStart = newStartBlock * BlockSize;
+                _availableStart = newStartBlock * BlockSize;
+            }
+        }
+
+        public void Commit(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= AvailableMemory.Length, $"MultiArrayBuffer.Commit: Expected byteCount={byteCount} <= {AvailableMemory.Length}");
+
+            _availableStart += byteCount;
+        }
+
+        // Ensure at least [byteCount] bytes to write to, up to the specified limit
+        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
+        {
+            if (ActiveMemory.Length >= limit)
+            {
+                // Already past limit. Do nothing.
+                return;
+            }
+
+            byteCount = Math.Min(byteCount, limit - ActiveMemory.Length);
+            if (byteCount <= AvailableMemory.Length)
+            {
+                // We have enough space available already.
+                return;
+            }
+
+            int newBytesNeeded = byteCount - AvailableMemory.Length;
+            int newBlocksNeeded = (newBytesNeeded + BlockSize - 1) / BlockSize;
+            Debug.Assert(newBlocksNeeded > 0);
+
+            if (_blocks is null)
+            {
+                Debug.Assert(_blockCount == 0);
+                Debug.Assert(_activeStart == 0);
+                Debug.Assert(_availableStart == 0);
+
+                int blockArraySize = 4;
+                while (blockArraySize < newBlocksNeeded)
+                {
+                    blockArraySize *= 2;
+                }
+
+                _blocks = new byte[]?[blockArraySize];
```

Comment:
Is it worth renting this from a pool, too?

Reply:
@jaredpar, I still struggle with our having reversed for nullability the positions of the array specifiers here :(  It's so confusing.

Reply:
Possibly. I'm not sure how well the pool does with small allocations, any thoughts?

The good news is that there will hopefully be significant reuse here if you have lots of MultiArrayBuffers, since we start with 4 entries here and then double from there. 

The bad news is that it seems pretty unlikely that anyone else is pooling arrays of byte[]...

Reply:
> Possibly. I'm not sure how well the pool does with small allocations, any thoughts?

Currently the smallest bucket size is 16, so any request for smaller will return a length 16 array.

Reply:
> The bad news is that it seems pretty unlikely that anyone else is pooling arrays of byte[]

No other code in the core libraries is.